src/Parser/Tokenizer/Tokenizer.php

```diff
@@ -127,25 +127,26 @@ private static function string(\Iterator $fragments): \Iterator
         $fragments->next();
 
         $buffer = Buffer::empty();
-        $escape = false;
 
         while ($fragments->valid()) {
-            /** @var Fragment $fragment */
-            $fragment = $fragments->current();
+            switch ($fragments->current()->value) {
+                case $delimiter:
+                    yield from $buffer->flush(TokenType::STRING_QUOTED);
+                    $fragments->next();
+                    return;
 
-            if (!$escape && $fragment->value === $delimiter) {
-                yield from $buffer->flush(TokenType::STRING_QUOTED);
-                $fragments->next();
-                return;
-            }
+                case '\\':
+                    $buffer->append($fragments->current());
+                    $fragments->next();
 
-            $escape = $fragment->value === '\\';
+                    $buffer->append($fragments->current());
```

Comment:
This breaks, unless `$fragments->valid()` is checked first.

Reply:
and this should imo break in case one does such cruel thing or not?

```
"hello world \
```
is not okay ...

```
                    if (!$fragments->valid()) {
                        throw new \Exception("@TODO: Unexpected end of input");
                    }
```